## SCAN Framework - Exploratory Data Analysis

| SCAN Framework | California Property Close Price Prediction |
|---|---|
| **Stakeholder Goal** | Understand California single-family home sales and identify the major factors and data issues relevant to predicting `ClosePrice`. The eventual goal is to support the development of a reliable property price prediction model. |
| **Columns and Coverage** | Focus on `ClosePrice`, property characteristics such as `LivingArea`, bedrooms, bathrooms, lot size, and year built, along with geographic and time variables. Analyze at least 6 months of `Residential` → `SingleFamilyResidence` sales data. |
| **Aggregates and Anomalies** | Examine typical `ClosePrice` and property characteristics using counts, distributions, and summary statistics. Identify missing values, duplicates, impossible or suspicious values, extreme prices, and other potential outliers that could affect modeling. |
| **Notable Segments** | Compare `ClosePrice` across meaningful segments such as property characteristics, city/ZIP/MLS area, and month to identify patterns that may be useful for prediction. |

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
oct_2025 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202510.csv")
nov_2025 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202511.csv")
dec_2025 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202512.csv")
jan_2026 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202601.csv")
feb_2026 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202602.csv")
march_2026 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202603.csv")
apr_2026 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202604.csv")


/var/folders/f7/b4dngbzs7y583pd4kd8nkp0w0000gp/T/ipykernel_96287/2364992055.py:4: DtypeWarning: Columns (0: WaterfrontYN, 1: PostalCode) have mixed types. Specify dtype option on import or set low_memory=False.
  jan_2026 = pd.read_csv("/Users/shouryakulkarni/IDX-Exchange-Data-Science/CRMLSSold202601.csv")


In [3]:
oct_2025['DataMonth'] = '2025-10'
nov_2025['DataMonth'] = '2025-11'
dec_2025['DataMonth'] = '2025-12'
jan_2026['DataMonth'] = '2026-01'
feb_2026['DataMonth'] = '2026-02'
march_2026['DataMonth'] = '2026-03'
apr_2026['DataMonth'] = '2026-04'

In [6]:
df = pd.concat([oct_2025,nov_2025, dec_2025, jan_2026, feb_2026, march_2026, apr_2026], ignore_index=True)
df

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,...,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,DataMonth,OriginatingSystemName,OriginatingSystemSubName
0,NorthSanLuisObispo,NorthSanLuisObispo,NaN,NaN,NaN,NaN,NaN,NaN,479293596,newlin@pacificaCRE.com,...,NaN,NaN,NaN,93446,NaN,1807.0,NaN,2025-10,NaN,NaN
1,TheInlandGateway,TheInlandGateway,NaN,True,NaN,NaN,NaN,12000.0,446914808,nleimers@hotmail.com,...,False,NaN,NaN,92325,0.0,14253.0,NaN,2025-10,NaN,NaN
2,SanDiego,SanDiego,Wood,True,NaN,NaN,False,1695.0,445176588,mannybehar@yahoo.com,...,NaN,0.0,NaN,92126,0.0,NaN,NaN,2025-10,NaN,NaN
3,SanDiego,SanDiego,NaN,False,NaN,NaN,False,950000.0,1145282039,chase@cromwellhomegroup.com,...,False,2.0,NaN,91901,NaN,NaN,NaN,2025-10,NaN,NaN
4,SanDiego,SanDiego,NaN,False,NaN,NaN,False,970000.0,1145278097,rory@corneliusestates.net,...,False,2.0,NaN,92115,NaN,NaN,NaN,2025-10,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145984,SierraNorthValley,SierraNorthValley,NaN,True,NaN,NaN,NaN,200000.0,1014741959,dtownsendrealestate@gmail.com,...,False,NaN,NaN,95916,0.0,910839.6,NaN,2026-04,CRMLS,CRMLS_CRM
145985,SouthwestRiversideCounty,InlandValleys,NaN,True,NaN,NaN,NaN,97500.0,1002540286,scott@westcoe.com,...,False,NaN,NaN,92539,0.0,522720.0,NaN,2026-04,CRMLS,CRMLS_CRM
145986,Mrmls,Southland,NaN,True,NaN,NaN,NaN,15999.0,482785043,mr.khatibi@yahoo.com,...,False,NaN,NaN,93551,0.0,20009.0,NaN,2026-04,CRMLS,CRMLS_CRM
145987,OrangeCounty,HighDesert,NaN,True,NaN,NaN,NaN,6000000.0,446619807,jeffnyal@gmail.com,...,False,NaN,NaN,92882,NaN,122839.0,NaN,2026-04,CRMLS,CRMLS_CRM


In [7]:
df.shape

(145989, 81)

In [84]:
df.columns

Index(['BuyerAgentAOR', 'ListAgentAOR', 'Flooring', 'ViewYN', 'WaterfrontYN',
       'BasementYN', 'PoolPrivateYN', 'OriginalListPrice', 'ListingKey',
       'ListAgentEmail', 'CloseDate', 'ClosePrice', 'ListAgentFirstName',
       'ListAgentLastName', 'Latitude', 'Longitude', 'UnparsedAddress',
       'PropertyType', 'LivingArea', 'ListPrice', 'DaysOnMarket',
       'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName',
       'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName',
       'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName',
       'FireplacesTotal', 'AssociationFeeFrequency', 'AboveGradeFinishedArea',
       'ListingKeyNumeric', 'MLSAreaMajor', 'TaxAnnualAmount',
       'CountyOrParish', 'MlsStatus', 'ElementarySchool', 'AttachedGarageYN',
       'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres',
       'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'StreetNumberNumeric',
       'ListingId', 'BathroomsTotalInteger', 'City', '

In [8]:
df['DataMonth'].value_counts().sort_index()

DataMonth
2025-10    23233
2025-11    19088
2025-12    20538
2026-01    16487
2026-02    19010
2026-03    23372
2026-04    24261
Name: count, dtype: int64

In [30]:
df['CloseDate'] = pd.to_datetime(df['CloseDate'], errors='coerce')

In [31]:
df = df[(df['PropertyType'] == 'Residential') & (df['PropertySubType'] == 'SingleFamilyResidence')].copy()
df

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,...,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,DataMonth,OriginatingSystemName,OriginatingSystemSubName
3,SanDiego,SanDiego,NaN,False,NaN,NaN,False,950000.0,1145282039,chase@cromwellhomegroup.com,...,False,2.0,NaN,91901,NaN,NaN,NaN,2025-10,NaN,NaN
4,SanDiego,SanDiego,NaN,False,NaN,NaN,False,970000.0,1145278097,rory@corneliusestates.net,...,False,2.0,NaN,92115,NaN,NaN,NaN,2025-10,NaN,NaN
5,SanDiego,SanDiego,NaN,False,NaN,NaN,False,575000.0,1145277224,garylawrence@kw.com,...,False,2.0,NaN,91945,NaN,NaN,NaN,2025-10,NaN,NaN
7,BeverlyHillsGreaterLa,BeverlyHillsGreaterLa,Tile,True,NaN,NaN,False,1243810.0,1145272633,taylor@acme-re.com,...,True,NaN,NaN,90065,141.25,2066.0,NaN,2025-10,NaN,NaN
8,BeverlyHillsGreaterLa,BeverlyHillsGreaterLa,Tile,True,NaN,NaN,False,1227060.0,1145272243,taylor@acme-re.com,...,True,NaN,NaN,90065,141.25,1804.0,NaN,2025-10,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
145955,CitrusValley,CitrusValley,"Carpet,Tile,Vinyl",True,NaN,NaN,False,649800.0,1073242811,Angelina@DannyGomesHomes.com,...,False,2.0,Los Angeles Unified,91331,0.00,7971.0,NaN,2026-04,CRMLS,CRMLS_CRM
145956,PacificWest,PasadenaFoothills,NaN,False,NaN,NaN,False,459000.0,1073214872,slcarter320@aol.com,...,False,2.0,Victor Valley Unified,92392,0.00,5200.0,NaN,2026-04,CRMLS,CRMLS_CRM
145957,LakeCounty,LakeCounty,NaN,True,NaN,True,False,279000.0,1069708148,jessica.homesells@gmail.com,...,False,0.0,Konocti Unified,95424,0.00,10019.0,NaN,2026-04,CRMLS,CRMLS_CRM
145959,TheInlandGateway,TheInlandGateway,"Carpet,Tile,Wood",False,NaN,True,False,449000.0,1067010583,brian@cohen-realty.com,...,False,0.0,Rim of the World,92322,0.00,12096.0,NaN,2026-04,CRMLS,CRMLS_CRM


In [32]:
df['DataMonth'].value_counts().sort_index()

DataMonth
2025-10    12029
2025-11     9739
2025-12    10455
2026-01     7490
2026-02     8959
2026-03    11562
2026-04    12490
Name: count, dtype: int64

In [33]:
df.columns

Index(['BuyerAgentAOR', 'ListAgentAOR', 'Flooring', 'ViewYN', 'WaterfrontYN',
       'BasementYN', 'PoolPrivateYN', 'OriginalListPrice', 'ListingKey',
       'ListAgentEmail', 'CloseDate', 'ClosePrice', 'ListAgentFirstName',
       'ListAgentLastName', 'Latitude', 'Longitude', 'UnparsedAddress',
       'PropertyType', 'LivingArea', 'ListPrice', 'DaysOnMarket',
       'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName',
       'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName',
       'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName',
       'FireplacesTotal', 'AssociationFeeFrequency', 'AboveGradeFinishedArea',
       'ListingKeyNumeric', 'MLSAreaMajor', 'TaxAnnualAmount',
       'CountyOrParish', 'MlsStatus', 'ElementarySchool', 'AttachedGarageYN',
       'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres',
       'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'StreetNumberNumeric',
       'ListingId', 'BathroomsTotalInteger', 'City', '

In [61]:
key_cols = [
    'ClosePrice',
    'LivingArea',
    'BedroomsTotal',
    'BathroomsTotalInteger',
    'LotSizeSquareFeet',
    'YearBuilt',
    'City',
    'PostalCode',
    'MLSAreaMajor',
    'CloseDate',
    'DataMonth', 'ClosePrice'
]

column_overview = pd.DataFrame({
    'dtype': df[key_cols].dtypes,
    'non_missing': df[key_cols].notna().sum(),
    'missing': df[key_cols].isna().sum(),
    'percent_missing': df[key_cols].isna().mean() * 100,
    'unique_values': df[key_cols].nunique()
})

column_overview



,dtype,non_missing,missing,percent_missing,unique_values
ClosePrice,float64,72724,0,0.000000,8071
LivingArea,float64,72690,34,0.046752,5225
BedroomsTotal,float64,72724,0,0.000000,18
BathroomsTotalInteger,float64,72723,1,0.001375,18
LotSizeSquareFeet,float64,71453,1271,1.747704,15738
YearBuilt,float64,72681,43,0.059128,149
City,str,72717,7,0.009625,920
PostalCode,object,72724,0,0.000000,1718
MLSAreaMajor,str,62298,10426,14.336395,990
CloseDate,datetime64[us],72724,0,0.000000,210


In [62]:
property_col_values = [
    'LivingArea',
    'BedroomsTotal',
    'BathroomsTotalInteger',
    'LotSizeSquareFeet',
    'YearBuilt', 'ClosePrice'
]

df[property_col_values].info()
df[property_col_values].describe()

<class 'pandas.DataFrame'>
Index: 72724 entries, 3 to 145962
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   LivingArea             72690 non-null  float64
 1   BedroomsTotal          72724 non-null  float64
 2   BathroomsTotalInteger  72723 non-null  float64
 3   LotSizeSquareFeet      71453 non-null  float64
 4   YearBuilt              72681 non-null  float64
 5   ClosePrice             72724 non-null  float64
dtypes: float64(6)
memory usage: 3.9 MB


,LivingArea,BedroomsTotal,BathroomsTotalInteger,LotSizeSquareFeet,YearBuilt,ClosePrice
count,72690.000000,72724.000000,72723.000000,7.145300e+04,72681.000000,7.272400e+04
mean,2050.484205,3.494885,2.636745,5.393820e+05,1976.118256,1.322379e+06
std,1036.467420,0.968269,1.130804,2.160306e+07,27.725184,7.424377e+06
min,0.000000,0.000000,0.000000,0.000000e+00,1801.000000,1.750000e+00
25%,1386.000000,3.000000,2.000000,5.663000e+03,1956.000000,6.190000e+05
50%,1820.000000,3.000000,2.000000,7.260000e+03,1977.000000,8.800000e+05
75%,2447.000000,4.000000,3.000000,1.045400e+04,1998.000000,1.400000e+06
max,31068.000000,22.000000,23.000000,1.938943e+09,2026.000000,9.895000e+08


In [63]:
df['City'].value_counts().head(10)


City
Los Angeles       2984
San Diego         2523
Riverside         1143
San Jose          1000
Oakland            941
Menifee            851
Long Beach         745
Lancaster          712
San Bernardino     659
Murrieta           656
Name: count, dtype: int64

In [64]:
df['PostalCode'].value_counts().head(10)


PostalCode
92253    560
92345    444
92584    411
92223    384
92211    381
92592    368
92596    348
92562    339
92336    329
93536    316
Name: count, dtype: int64

In [65]:
df['MLSAreaMajor'].value_counts().head(10)

MLSAreaMajor
699 - Not Defined                       6127
SRCAR - Southwest Riverside County      4055
252 - Riverside                         1151
LAC - Lancaster                          702
263 - Banning/Beaumont/Cherry Valley     648
274 - San Bernardino                     647
VIC - Victorville                        615
259 - Moreno Valley                      601
248 - Corona                             598
APPV - Apple Valley                      552
Name: count, dtype: int64

In [66]:
df.groupby('DataMonth').size()

DataMonth
2025-10    12029
2025-11     9739
2025-12    10455
2026-01     7490
2026-02     8959
2026-03    11562
2026-04    12490
dtype: int64

In [67]:
#Percentage of missing values for each column across all of the months
missing_by_month = (df.groupby('DataMonth')[key_cols].apply(lambda x: x.isna().mean() * 100))

missing_by_month

,ClosePrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,LotSizeSquareFeet,YearBuilt,City,PostalCode,MLSAreaMajor,CloseDate,DataMonth,ClosePrice
DataMonth,,,,,,,,,,,,
2025-10,0.0,0.041566,0.0,0.000000,1.787347,0.049879,0.000000,0.0,15.628897,0.0,0.0,0.0
2025-11,0.0,0.041072,0.0,0.000000,1.499127,0.051340,0.000000,0.0,14.929664,0.0,0.0,0.0
2025-12,0.0,0.047824,0.0,0.000000,1.492109,0.057389,0.009565,0.0,13.868962,0.0,0.0,0.0
2026-01,0.0,0.120160,0.0,0.000000,1.975968,0.093458,0.013351,0.0,12.563418,0.0,0.0,0.0
2026-02,0.0,0.033486,0.0,0.000000,2.009153,0.033486,0.044648,0.0,14.052908,0.0,0.0,0.0
2026-03,0.0,0.025947,0.0,0.008649,1.591420,0.051894,0.000000,0.0,13.907628,0.0,0.0,0.0
2026-04,0.0,0.040032,0.0,0.000000,1.937550,0.080064,0.008006,0.0,14.683747,0.0,0.0,0.0


### Columns & Coverage Findings

- The dataset covers seven months from October 2025 through April 2026.
- `ClosePrice`, `BedroomsTotal`, `PostalCode`, and `CloseDate` have complete coverage across the analyzed months.
- Core property characteristics such as `LivingArea`, bathrooms, and `YearBuilt` have very low missingness.
- `LotSizeSquareFeet` has approximately 1.5–2.0% missingness depending on the month.
- `MLSAreaMajor` has the highest missingness among the selected features, ranging from approximately 12.6–15.6%.
- Missingness is generally consistent across months, with no obvious major deterioration in data coverage over time.

In [70]:
numeric_cols = [
    'ClosePrice',
    'LivingArea',
    'BedroomsTotal',
    'BathroomsTotalInteger',
    'LotSizeSquareFeet',
    'YearBuilt'
]

df[numeric_cols].describe()

,ClosePrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,LotSizeSquareFeet,YearBuilt
count,7.272400e+04,72690.000000,72724.000000,72723.000000,7.145300e+04,72681.000000
mean,1.322379e+06,2050.484205,3.494885,2.636745,5.393820e+05,1976.118256
std,7.424377e+06,1036.467420,0.968269,1.130804,2.160306e+07,27.725184
min,1.750000e+00,0.000000,0.000000,0.000000,0.000000e+00,1801.000000
25%,6.190000e+05,1386.000000,3.000000,2.000000,5.663000e+03,1956.000000
50%,8.800000e+05,1820.000000,3.000000,2.000000,7.260000e+03,1977.000000
75%,1.400000e+06,2447.000000,4.000000,3.000000,1.045400e+04,1998.000000
max,9.895000e+08,31068.000000,22.000000,23.000000,1.938943e+09,2026.000000


Distribution is right-skewed, since the mean `ClosePrice` is greater than the median `ClosePrice`

In [71]:
#investigating the properties with the 10 smallest ClosePrices
df.nsmallest(10, 'ClosePrice')[
    ['ClosePrice', 'LivingArea', 'BedroomsTotal',
     'BathroomsTotalInteger', 'City', 'DataMonth']
]

,ClosePrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,City,DataMonth
73450,1.75,3513.0,3.0,4.0,Palm Desert,2026-01
10330,485.00,1305.0,3.0,2.0,Cathedral City,2025-10
90706,8000.00,984.0,3.0,1.0,Trona,2026-02
78946,8300.00,3886.0,5.0,5.0,Altadena,2026-01
98089,18000.00,1488.0,0.0,0.0,Blythe,2026-02
111453,28000.00,408.0,0.0,0.0,Landers,2026-03
37944,29000.00,960.0,3.0,1.0,Yermo,2025-11
23534,29900.00,1198.0,1.0,1.0,Buttonwillow,2025-11
47908,31000.00,1942.0,5.0,3.0,Trona,2025-12
74526,31000.00,1309.0,4.0,2.0,Bakersfield,2026-01


In [72]:
#investigating the properties with the 10 largest ClosePrices
df.nlargest(10, 'ClosePrice')[
    ['ClosePrice', 'LivingArea', 'BedroomsTotal',
     'BathroomsTotalInteger', 'City', 'DataMonth']
]


,ClosePrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,City,DataMonth
5842,989500000.0,2289.0,5.0,3.0,Valencia,2025-10
142178,796000000.0,1936.0,2.0,3.0,Valley Center,2026-04
141262,768500000.0,1444.0,4.0,2.0,San Diego,2026-04
37463,740000000.0,1271.0,3.0,2.0,San Diego,2025-11
115953,664250000.0,1992.0,4.0,3.0,Chula Vista,2026-03
135035,645000000.0,1104.0,2.0,1.0,Oceanside,2026-04
75786,472000000.0,1910.0,4.0,2.0,Victorville,2026-01
41504,199947850.0,1704.0,3.0,2.0,Topanga,2025-11
127127,60000000.0,15006.0,7.0,14.0,Montecito,2026-04
23222,53000000.0,7865.0,3.0,5.0,Bonsall,2025-10


In [73]:
#investigating the properties with the 10 largest LotSizeSquareFeet
df.nlargest(10, 'LotSizeSquareFeet')[
    ['LotSizeSquareFeet', 'LivingArea', 'BedroomsTotal',
     'BathroomsTotalInteger', 'City', 'DataMonth']
]


,LotSizeSquareFeet,LivingArea,BedroomsTotal,BathroomsTotalInteger,City,DataMonth
26286,1.938943e+09,2825.0,4.0,4.0,Agoura,2025-11
95795,1.897474e+09,3549.0,5.0,4.0,Orange,2026-02
118906,1.897474e+09,1990.0,3.0,3.0,Mountain Center,2026-03
76048,1.802600e+09,3035.0,5.0,3.0,Madera,2026-01
25944,1.432688e+09,912.0,2.0,1.0,San Bernardino,2025-11
31953,1.306800e+09,3155.0,3.0,4.0,Apple Valley,2025-11
27191,1.123848e+09,2840.0,3.0,3.0,Fresno,2025-11
74774,1.058334e+09,1260.0,3.0,3.0,Valencia,2026-01
75150,1.053281e+09,1372.0,3.0,2.0,Apple Valley,2026-01
59215,1.047095e+09,2549.0,4.0,3.0,Apple Valley,2025-12


In [75]:
#investigating the properties with the 10 largest LivingAreas

df.nlargest(10, 'LivingArea')[
    ['LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger',
     'ClosePrice', 'City', 'DataMonth']
]

,LivingArea,BedroomsTotal,BathroomsTotalInteger,ClosePrice,City,DataMonth
41738,31068.0,19.0,23.0,28000000.0,Rancho Santa Fe,2025-11
107696,19474.0,7.0,15.0,30260000.0,Beverly Hills,2026-03
96633,17153.0,10.0,13.0,8000000.0,Bradbury,2026-02
22787,16683.0,8.0,12.0,11515000.0,La Jolla,2025-10
117842,16499.0,9.0,13.0,3500000.0,Los Angeles,2026-03
120769,16489.0,6.0,12.0,15300000.0,Malibu,2026-03
23040,15355.0,9.0,14.0,32000000.0,Santa Barbara,2025-10
79304,15219.0,7.0,13.0,31500000.0,Newport Coast,2026-01
79725,15037.0,8.0,12.0,10300000.0,Rancho Santa Fe,2026-02
63036,15015.0,7.0,9.0,16426180.0,Encino,2026-01


In [ ]:
#investigating the properties with the 10 largest number of BedroomsTotal
df.nlargest(10, 'BedroomsTotal')[
    ['BedroomsTotal', 'LivingArea', 'BathroomsTotalInteger',
     'ClosePrice', 'City', 'DataMonth']
]



,BedroomsTotal,LivingArea,BathroomsTotalInteger,ClosePrice,City,DataMonth
67062,22.0,5814.0,18.0,1200000.0,Los Angeles,2026-01
41738,19.0,31068.0,23.0,28000000.0,Rancho Santa Fe,2025-11
62361,16.0,8305.0,11.0,2555000.0,Granada Hills,2025-12
38866,15.0,1945.0,2.0,319000.0,Bakersfield,2025-11
29289,13.0,4959.0,5.0,800000.0,Los Angeles,2025-11
59268,12.0,7500.0,12.0,1500000.0,San Bernardino,2025-12
79149,12.0,8487.0,10.0,4600000.0,Woodland Hills,2026-01
16553,11.0,2946.0,5.0,1125000.0,Reseda,2025-10
56920,11.0,4480.0,5.0,1175000.0,Los Angeles,2025-12
92632,11.0,2989.0,6.0,1540000.0,San Diego,2026-02


In [ ]:
#investigating the properties with the 10 largest number of BathroomsTotalInteger
df.nlargest(10, 'BathroomsTotalInteger')[
    ['BathroomsTotalInteger', 'LivingArea', 'BedroomsTotal',
     'ClosePrice', 'City', 'DataMonth']
    
]


,BathroomsTotalInteger,LivingArea,BedroomsTotal,ClosePrice,City,DataMonth
41738,23.0,31068.0,19.0,28000000.0,Rancho Santa Fe,2025-11
67062,18.0,5814.0,22.0,1200000.0,Los Angeles,2026-01
107696,15.0,19474.0,7.0,30260000.0,Beverly Hills,2026-03
23040,14.0,15355.0,9.0,32000000.0,Santa Barbara,2025-10
127127,14.0,15006.0,7.0,60000000.0,Montecito,2026-04
145160,14.0,11316.0,7.0,16200000.0,La Quinta,2026-04
20670,13.0,12664.0,9.0,34000000.0,Beverly Hills,2025-10
62306,13.0,13404.0,9.0,23102000.0,Beverly Hills,2025-12
79304,13.0,15219.0,7.0,31500000.0,Newport Coast,2026-01
96633,13.0,17153.0,10.0,8000000.0,Bradbury,2026-02


In [ ]:
#investigating 10 of the oldest properties across the dataset by YearBuilt
df.nsmallest(10, 'YearBuilt')[
    ['YearBuilt', 'City', 'LivingArea', 'ClosePrice', 'DataMonth']
]
#the Berkeley property has the smallest living area, along with Needles

,YearBuilt,City,LivingArea,ClosePrice,DataMonth
52424,1801.0,Needles,930.0,123000.0,2025-12
94272,1801.0,San Bernardino,1221.0,310000.0,2026-02
121316,1801.0,Needles,898.0,44000.0,2026-03
73509,1854.0,Alameda,2750.0,1390000.0,2026-01
37839,1872.0,Santa Cruz,1923.0,1270000.0,2025-11
61861,1872.0,Lewiston,2700.0,1193750.0,2025-12
46305,1876.0,Pacific Grove,1295.0,2000000.0,2025-12
35163,1878.0,Berkeley,704.0,750000.0,2025-11
3869,1880.0,Alameda,2414.0,1930000.0,2025-10
54506,1880.0,Mendocino,1380.0,1228160.0,2025-12


In [83]:
#Investigating LotSizes for outliers
lot_outliers = df.nlargest(10, 'LotSizeSquareFeet')

lot_outliers[[
    'LotSizeSquareFeet',
    'LotSizeArea',
    'LotSizeAcres',
    'LotSizeDimensions',
    'LivingArea',
    'City',
    'DataMonth'
]]

,LotSizeSquareFeet,LotSizeArea,LotSizeAcres,LotSizeDimensions,LivingArea,City,DataMonth
26286,1.938943e+09,44512.0,44512.0,NaN,2825.0,Agoura,2025-11
95795,1.897474e+09,43560.0,43560.0,NaN,3549.0,Orange,2026-02
118906,1.897474e+09,43560.0,43560.0,NaN,1990.0,Mountain Center,2026-03
76048,1.802600e+09,41382.0,41382.0,NaN,3035.0,Madera,2026-01
25944,1.432688e+09,32890.0,32890.0,NaN,912.0,San Bernardino,2025-11
31953,1.306800e+09,30000.0,30000.0,30000,3155.0,Apple Valley,2025-11
27191,1.123848e+09,25800.0,25800.0,NaN,2840.0,Fresno,2025-11
74774,1.058334e+09,24296.0,24296.0,NaN,1260.0,Valencia,2026-01
75150,1.053281e+09,24180.0,24180.0,NaN,1372.0,Apple Valley,2026-01
59215,1.047095e+09,24038.0,24038.0,NaN,2549.0,Apple Valley,2025-12


In [86]:
# Check whether LotSizeSquareFeet = LotSizeAcres * 43,560
# Only rows where both lot-size fields are available
lot_check = df[['LotSizeAcres', 'LotSizeSquareFeet']].dropna().copy()

lot_check['ExpectedLotSqFt'] = lot_check['LotSizeAcres'] * 43560

np.isclose(
    lot_check['LotSizeSquareFeet'],
    lot_check['ExpectedLotSqFt']
).mean()

np.float64(0.14890549770462433)

Lot-size data quality: Extremely large LotSizeSquareFeet values were identified. While several extreme observations are consistent with LotSizeAcres × 43,560, only approximately 14.9% of non-missing records satisfy this relationship overall. Because LotSizeUnits is not available in the provided dataset, the inconsistency cannot be fully resolved during EDA. Lot-size fields will therefore require further validation during preprocessing.

In [ ]:
#Number of sales, median close price, and median living area by month
monthly_summary = df.groupby('DataMonth').agg(
    number_of_sales=('ClosePrice', 'count'),
    median_close_price=('ClosePrice', 'median'),
    median_living_area=('LivingArea', 'median')
)

monthly_summary

,number_of_sales,median_close_price,median_living_area
DataMonth,,,
2025-10,12029,888163.0,1824.0
2025-11,9739,875000.0,1812.0
2025-12,10455,860000.0,1804.0
2026-01,7490,851691.5,1821.0
2026-02,8959,875000.0,1823.5
2026-03,11562,890000.0,1821.0
2026-04,12490,915000.0,1833.0


In [ ]:
#Number of sales, mean close price, median close price, and median living area by city

city_summary = df.groupby('City').agg(
    number_of_sales=('ClosePrice', 'count'),
    mean_close_price=('ClosePrice', 'mean'),
    median_close_price=('ClosePrice', 'median'),
    median_living_area=('LivingArea', 'median')
).sort_values('number_of_sales', ascending=False)
city_summary

,number_of_sales,mean_close_price,median_close_price,median_living_area
City,,,,
Los Angeles,2984,1.924095e+06,1370000.0,1709.0
San Diego,2523,2.000950e+06,1189000.0,1671.0
Riverside,1143,7.425922e+05,678000.0,1773.0
San Jose,1000,1.762196e+06,1617500.0,1633.0
Oakland,941,1.083639e+06,925000.0,1545.0
...,...,...,...,...
Del Cerro,1,1.285000e+06,1285000.0,2303.0
Klamath River,1,2.350000e+05,235000.0,750.0
Knightsen,1,1.375000e+06,1375000.0,1356.0


In [91]:
#Number of sales, mean close price, median close price, and median living area by for cities with more than 100 sales
city_100_summary = df.groupby('City').agg(
    number_of_sales=('ClosePrice', 'count'),
    mean_close_price=('ClosePrice', 'mean'),
    median_close_price=('ClosePrice', 'median'),
    median_living_area=('LivingArea', 'median')
).sort_values('number_of_sales', ascending=False)

city_100_summary = city_100_summary[city_100_summary['number_of_sales'] > 100]
city_100_summary

,number_of_sales,mean_close_price,median_close_price,median_living_area
City,,,,
Los Angeles,2984,1.924095e+06,1370000.0,1709.0
San Diego,2523,2.000950e+06,1189000.0,1671.0
Riverside,1143,7.425922e+05,678000.0,1773.0
San Jose,1000,1.762196e+06,1617500.0,1633.0
Oakland,941,1.083639e+06,925000.0,1545.0
...,...,...,...,...
Rowland Heights,102,1.132791e+06,994944.0,1965.5
Indian Wells,101,2.678848e+06,2175000.0,3260.0
Rancho Santa Fe,101,5.595726e+06,4773000.0,5355.0
